File: algorithms/632. Smallest Range/solution.cc
class Solution {
public:
    vector<int> smallestRange(vector<vector<int>>& nums) {
        // @tmt514
        vector<int> pos(nums.size());
        set<pair<int, int>> q;
        for ( int i = 0, n = nums.size(); i < n; ++i )
            q.emplace(make_pair(nums[i][pos[i]++], i));
        int left = q.begin()->first, right = q.rbegin()->first;
        while ( true ) {
            int val = q.begin()->first, i = q.begin()->second;
            if ( val == nums[i].back() )
                break;
            q.erase(q.begin());
            q.emplace(make_pair(nums[i][pos[i]++], i));
            int l = q.begin()->first, r = q.rbegin()->first;
            if ( right - left > r - l ) {
                left = l;
                right = r;
            }
        }
        return {left, right};
    }
};

// 86 / 86 test cases passed.
// Status: Accepted
// Runtime: 46 ms
// beats 67.50 %

File: algorithms/632. Smallest Range/solution.py
class Solution(object):
    def smallestRange(self, nums):
        """
        :type nums: List[List[int]]
        :rtype: List[int]
        """
        h = [[a[0], i, 0] for i, a in enumerate(nums)]
        heapq.heapify(h)
        result = [float('-inf'), float('inf')]
        right = max(a[0] for a in nums)
        while True:
            left, i, j = heapq.heappop(h)
            if right - left < result[1] - result[0]:
                result = [left, right]
            if j + 1 == len(nums[i]):
                return result
            right = max(right, nums[i][j+1])
            heapq.heappush(h, [nums[i][j+1], i, j+1])
